src/emitter.ts

```diff
@@ -0,0 +1,1126 @@
+        let returnType = convertDomTypeToTsType(s);
+        returnType = s.nullable ? makeNullable(returnType) : returnType;
+        printLine(`${prefix || ""}${name || ""}(${paramsString}): ${returnType};`);
+    }
+
+    function emitSignatures(method: { signature?: Browser.Signature[], "override-signatures"?: string[], "additional-signatures"?: string[] }, prefix: string, name: string, printLine: (s: string) => void) {
+        if (method["override-signatures"]) {
+            method["override-signatures"]!.forEach(s => printLine(`${prefix}${s};`));
+        }
+        else if (method.signature) {
+            if (method["additional-signatures"]) {
+                method["additional-signatures"]!.forEach(s => printLine(`${prefix}${s};`));
+            }
+            method.signature.forEach(sig => emitSignature(sig, prefix, name, printLine));
+        }
+    }
+
+    function emitMethods(prefix: string, emitScope: EmitScope, i: Browser.Interface, conflictedMembers: Set<string>) {
+        // If prefix is not empty, then this is the global declare function addEventListener, we want to override this
+        // Otherwise, this is EventTarget.addEventListener, we want to keep that.
+        function mFilter(m: Browser.Method) {
+            return matchScope(emitScope, m) &&
+                !(prefix !== "" && (m.name === "addEventListener" || m.name === "removeEventListener"));
+        }
+
+        if (i.methods) {
+            mapToArray(i.methods.method)
+                .filter(mFilter)
+                .sort(compareName)
+                .forEach(m => emitMethod(prefix, i, m, conflictedMembers));
+        }
+
+        // The window interface inherited some methods from "Object",
+        // which need to explicitly exposed
+        if (i.name === "Window" && prefix === "declare function ") {
+            printer.printLine("declare function toString(): string;");
+        }
+    }
+
+    /// Emit the properties and methods of a given interface
+    function emitMembers(prefix: string, emitScope: EmitScope, i: Browser.Interface) {
+        const conflictedMembers = extendConflictsBaseTypes[i.name] ? extendConflictsBaseTypes[i.name].memberNames : new Set();
+        emitProperties(prefix, emitScope, i, conflictedMembers);
+        const methodPrefix = prefix.startsWith("declare var") ? "declare function " : "";
+        emitMethods(methodPrefix, emitScope, i, conflictedMembers);
+    }
+
+    /// Emit all members of every interfaces at the root level.
+    /// Called only once on the global polluter object
+    function emitAllMembers(i: Browser.Interface) {
+        const prefix = "declare var ";
+        emitMembers(prefix, EmitScope.All, i);
+
+        for (const relatedIName of iNameToIDependList[i.name]) {
+            const i = allInterfacesMap[relatedIName];
+            if (i) {
+                emitAllMembers(i);
+            }
+        }
+    }
+
+    function emitEventHandlers(prefix: string, i: Browser.Interface) {
```

Comment:
```ts
function emitEventHandlers(i: Browser.Interface, inNamespace = false) {
    for (const addOrRemove of ["add", "remove"]) {
        const optionsType = addOrRemove === "add" ? "AddEventListenerOptions" : "EventListenerOptions";
        const fPrefix = inNamespace ? "declare function " : "";
        if (tryEmitTypedEventHandlerForInterface()) {
            // only emit the string event handler if we just emited a typed handler
            printer.printLine(`${fPrefix}${addOrRemove}EventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | ${optionsType}): void;`);
        }

        function emitTypedEventHandler(iParent: Browser.Interface) {
            printer.printLine(`${fPrefix}${addOrRemove}EventListener<K extends keyof ${iParent.name}EventMap>(type: K, listener: (this: ${i.name}, ev: ${iParent.name}EventMap[K]) => any, options?: boolean | ${optionsType}): void;`);
        }

        function tryEmitTypedEventHandlerForInterface() {
            if (iNameToEhList[i.name] && iNameToEhList[i.name].length) {
                emitTypedEventHandler(i);
                return true;
            }
            if (iNameToEhParents[i.name] && iNameToEhParents[i.name].length) {
                iNameToEhParents[i.name]
                    .sort(compareName)
                    .forEach(emitTypedEventHandler);
                return true;
            }
            return false;
        }
    }
}
```

Reply:
not sure how i feel about functions declared in a for loop. will refactor though.